File: apps/visualizer/backend/api/schemas/scores.py
"""Scores API response models — single source of truth for image_scores shapes."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class ImageScoreRow(BaseModel):
    model_config = ConfigDict(extra='forbid')

    id: int | None = None
    image_key: str
    image_type: str
    perspective_slug: str
    score: int
    rationale: str = ''
    model_used: str = ''
    prompt_version: str = ''
    scored_at: str
    is_current: bool
    repaired_from_malformed: bool = False
    not_attempted: bool = False


class ScoresCurrentResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    image_key: str
    image_type: str
    current: list[ImageScoreRow]


class ScoresHistoryResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    image_key: str
    image_type: str
    perspective_slug: str
    history: list[ImageScoreRow]


def validate_image_score_row(row: dict) -> dict:
    """Validate a normalized score row; raises on shape drift."""
    return ImageScoreRow.model_validate(row).model_dump(mode='json')

File: apps/visualizer/backend-ts/src/api/schemas/scores.ts
/**
 * Scores API response models — single source of truth for `image_scores` shapes.
 * Port of `api/schemas/scores.py`.
 */
import { z } from '@hono/zod-openapi';

export const ImageScoreRow = z
  .object({
    id: z.int().nullish(),
    image_key: z.string(),
    image_type: z.string(),
    perspective_slug: z.string(),
    score: z.int(),
    rationale: z.string().nullish(),
    model_used: z.string().nullish(),
    prompt_version: z.string(),
    scored_at: z.string(),
    is_current: z.boolean(),
    repaired_from_malformed: z.boolean(),
    not_attempted: z.boolean().default(false),
  })
  .strict()
  .openapi('ImageScoreRow');

export const ScoresCurrentResponse = z
  .object({
    image_key: z.string(),
    image_type: z.literal('catalog'),
    current: z.array(ImageScoreRow),
  })
  .strict()
  .openapi('ScoresCurrentResponse');

export const ScoresHistoryResponse = z
  .object({
    image_key: z.string(),
    image_type: z.literal('catalog'),
    perspective_slug: z.string(),
    history: z.array(ImageScoreRow),
  })
  .strict()
  .openapi('ScoresHistoryResponse');

export type ImageScoreRow = z.infer<typeof ImageScoreRow>;
